filewatcher/v2/filewatcher.go

```diff
@@ -0,0 +1,567 @@
+	return r.data.Load().(*atomicData[T]).data
+}
+
+// Close stops the FileWatcher.
+//
+// After Close is called you won't get any updates on the file content,
+// but you can still call Get to get the last content before stopping.
+//
+// It's OK to call Close multiple times.
+// Calls after the first one are essentially no-op.
+//
+// Close never returns an error.
+func (r *Result[T]) Close() error {
+	r.cancel()
+	return nil
+}
+
+func getMtime(path string) (time.Time, error) {
+	stat, err := os.Stat(path)
+	if err != nil {
+		return time.Time{}, err
+	}
+	return stat.ModTime(), nil
+}
+
+func (r *Result[T]) watcherLoop(
+	watcher *fsnotify.Watcher,
+	path string,
+	parser Parser[T],
+	softLimit, hardLimit int64,
+	pollingInterval time.Duration,
+	fsEventsDelay time.Duration,
+) {
+	var lock sync.Mutex
+	forceReload := func() {
+		// make sure we don't run forceReload concurrently
+		lock.Lock()
+		defer lock.Unlock()
+
+		d, mtime, files, err := openAndParse(path, parser, softLimit, hardLimit)
+		if err != nil {
+			slog.ErrorContext(r.ctx, "filewatcher: openAndParse returned error", "err", err)
+		} else {
+			r.data.Store(&atomicData[T]{
+				data:  d,
+				mtime: mtime,
+			})
+			// remove all previously watched files
+			for _, path := range watcher.WatchList() {
+				watcher.Remove(path)
+			}
+			// then read all new files to watch
+			for _, path := range files {
+				if err := watcher.Add(path); err != nil {
+					slog.ErrorContext(r.ctx, "filewatcher: failed to watch file", "err", err, "path", path)
+				}
+			}
+		}
+	}
+
+	reload := func() {
+		mtime, err := getMtime(path)
+		if err != nil {
+			slog.ErrorContext(r.ctx, "filewatcher: failed to get mtime for file", "err", err, "path", path)
+			return
+		}
+		if r.data.Load().(*atomicData[T]).mtime.Before(mtime) {
+			forceReload()
+		}
+	}
+
+	var tickerChan <-chan time.Time
+	if pollingInterval > 0 {
+		ticker := time.NewTicker(pollingInterval)
+		defer ticker.Stop()
+		tickerChan = ticker.C
+	}
+	var timer *time.Timer
+	for {
+		select {
+		case <-r.ctx.Done():
+			watcher.Close()
+			return
+
+		case err := <-watcher.Errors:
+			slog.ErrorContext(r.ctx, "filewatcher: watcher error", "err", err)
+
+		case ev := <-watcher.Events:
+			// When both r.ctx.Done() and watcher.Events are unblocked, there's no
+			// guarantee which case would be picked, so do an additional ctx check
+			// here to make sure we don't spam the log with i/o errors (which mainly
+			// happen in tests)
+			if r.ctx.Err() != nil {
+				continue
+			}
+
+			switch ev.Op {
+			default:
+				// Ignore uninterested events.
+			case fsnotify.Create, fsnotify.Write, fsnotify.Rename, fsnotify.Remove:
+				// Use fsEventDelay to avoid calling forceReload repetively when a burst
+				// of fs events happens (for example, when multiple files within the
+				// directory changed).
+				if timer == nil {
+					timer = time.AfterFunc(fsEventsDelay, forceReload)
+				} else {
+					timer.Reset(fsEventsDelay)
```

Comment:
reset is not safe to call if there is a chance that it might be expiring concurrently (which is basically always)

https://pkg.go.dev/time#Timer.Reset

```
if !t.Stop() {
	<-t.C
}
t.Reset(d)
```

Reply:
the timer here was created via `time.AfterFunc`, if you look at its doccomment, it says:
>The returned Timer's C field is not used and will be nil.

Reply:
Probably a good idea to add a comment about this, I could see someone reading this code later and being confused.

Reply:
I think the problem is that both `time.NewTimer` and `time.AfterFunc` creates a `*time.Timer`, but `time.Timer.Reset` has this caveat if it's created via `time.NewTimer` that you should do the additional check to avoid race condition, but that `Reset` itself is wrong (you just must always use it with some additional checks in certain conditions).

so overall I think the better approach is to be very careful to `time.NewTimer` instead of being careful to `time.Timer.Reset`. in most cases, you use `time.NewTimer` to do something after the timer fires, which is better done via `time.AfterFunc` which is much safer. or some other cases it might be able to be replaced by `time.NewTicker`. I'm not sure if there are some cases you should use `time.NewTimer` over the other 2 (maybe there are some very niche cases).

Reply:
added comments linking back to this thread.